File: 32-longest-valid-parentheses/longest-valid-parentheses.ts
function longestValidParentheses(s: string): number {
    let maxLength: number = 0
    const stack: number[] = [-1]

    for (let i = 0; i < s.length; i++) {
        if (s[i] === '(') {
            stack.push(i)
        } else {
            stack.pop()

            if (stack.length === 0) {
                stack.push(i)
            } else {
                maxLength = Math.max(maxLength, i - stack[stack.length - 1])
            }
        }
    }
    return maxLength
};

File: 32-longest-valid-parentheses/longest-valid-parentheses.py
class Solution:
    def longestValidParentheses(self, s: str) -> int:
        max_len = 0
        stack = [-1]
        
        for i, char in enumerate(s):
            if char == '(':
                stack.append(i)
            else:
                # first pop the top element
                stack.pop()
                if not stack:
                    # stack is empty. update the start of the valid substring
                    stack.append(i)
                else:
                    # stack is not empty.
                    # we can calculate the length of the valid substring
                    max_len = max(max_len, i - stack[-1])
        
        return max_len